pkg/app/server/apikeyverifier/verifier.go

```diff
@@ -23,24 +23,30 @@ import (
 
 	"github.com/pipe-cd/pipecd/pkg/cache"
 	"github.com/pipe-cd/pipecd/pkg/cache/memorycache"
+	"github.com/pipe-cd/pipecd/pkg/cache/rediscache"
 	"github.com/pipe-cd/pipecd/pkg/model"
+	"github.com/pipe-cd/pipecd/pkg/redis"
 )
 
 type apiKeyGetter interface {
 	Get(ctx context.Context, id string) (*model.APIKey, error)
 }
 
 type Verifier struct {
-	apiKeyCache cache.Cache
-	apiKeyStore apiKeyGetter
-	logger      *zap.Logger
+	apiKeyCache         cache.Cache
+	apiKeyStore         apiKeyGetter
+	apiKeyLastUsedCache cache.Cache
+	logger              *zap.Logger
 }
 
-func NewVerifier(ctx context.Context, getter apiKeyGetter, logger *zap.Logger) *Verifier {
+const apiKeyLastUsedCacheHashKey = "HASHKEY:PIPED:API_KEYS" //nolint:gosec
+
+func NewVerifier(ctx context.Context, getter apiKeyGetter, rd redis.Redis, logger *zap.Logger) *Verifier {
```

Comment:
instead of passing redis directly here and creating HashCache based redis in the constructor, you can create HashCache for lastUsedCache in server.go and pass it to this constructor.

https://github.com/pipe-cd/pipecd/blob/master/cmd/pipecd/server.go#L198-L213
https://github.com/pipe-cd/pipecd/blob/master/cmd/pipecd/server.go#L72

That also helps reduce the definition for `apiKeyLastUsedCacheHashKey` to one (only have to define it in server.go)

Reply:
Or maybe if we only need Putter, let's define interface Putter here just like we did with 
https://github.com/pipe-cd/pipecd/pull/4124/files#diff-b8762b28ce9988e536f2d9961930f127631c2e8a55e489963913dfbda93e6567R31